cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStoreProvider.java

```diff
@@ -0,0 +1,543 @@
+      } catch (IOException ioe) {
+        // rename _OLD back to current and throw Exception
+        revertFromOld(oldPath, fileExisted);
+        resetPath = path;
+        throw ioe;
+      }
+      // Rename _NEW to CURRENT and delete _OLD
+      cleanupNewAndOld(newPath, oldPath);
+      changed = false;
+    } catch (IOException ioe) {
+      resetKeyStoreState(resetPath);
+      LOG.error("Failed to persist the key store. Secure data may be lost on a restart.", ioe);
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  private boolean backupToOld(Path oldPath) throws IOException {
+    boolean fileExisted = false;
+    if (Files.exists(path)) {
+      renameOrFail(path, oldPath);
+      fileExisted = true;
+    }
+    return fileExisted;
+  }
+
+  private void resetKeyStoreState(Path path) {
+    LOG.debug("Could not flush Keystore.."
+                + "attempting to reset to previous state !!");
+    // 1) flush cache
+    cache.clear();
+    // 2) load keyStore from previous path
+    try {
+      loadFromPath(keyStore, path, password);
+      LOG.debug("KeyStore resetting to previously flushed state !!");
+    } catch (Exception e) {
+      LOG.debug("Could not reset Keystore to previous state", e);
+    }
+  }
+
+  private void cleanupNewAndOld(Path newPath, Path oldPath) throws IOException {
+    // Rename _NEW to CURRENT
+    renameOrFail(newPath, path);
+    // Delete _OLD
+    if (Files.exists(oldPath)) {
+      Files.delete(oldPath);
+    }
+  }
+
+  private void writeToNew(Path newPath) throws IOException {
+    try (OutputStream fos = new DataOutputStream(Files.newOutputStream(newPath))) {
+      keyStore.store(fos, password);
+    } catch (KeyStoreException e) {
+      throw new IOException("Can't store keystore " + this, e);
+    } catch (NoSuchAlgorithmException e) {
+      throw new IOException(
+        "No such algorithm storing keystore " + this, e);
+    } catch (CertificateException e) {
+      throw new IOException(
+        "Certificate exception storing keystore " + this, e);
+    }
+  }
+
+  private void revertFromOld(Path oldPath, boolean fileExisted)
+    throws IOException {
+    if (fileExisted) {
+      renameOrFail(oldPath, path);
+    }
+  }
+
+  /**
+   * An adapter between a KeyStore Key and our SecureStoreMetadata. This is used to store
+   * the metadata in a KeyStore even though isn't really a key.
+   */
+  private static class KeyMetadata implements Key, Serializable {
+    private SecureStoreMetadata metadata;
+    private static final long serialVersionUID = 3405839418917868651L;
+    private static final String METADATA_FORMAT = "KeyMetadata";
+
+    private KeyMetadata(SecureStoreMetadata meta) {
+      this.metadata = meta;
+    }
+
+    @Override
+    public String getAlgorithm() {
+      return ALGORITHM_PROXY;
+    }
+
+    @Override
+    public String getFormat() {
+      return METADATA_FORMAT;
+    }
+
+    @Override
+    public byte[] getEncoded() {
+      return new byte[0];
```

Comment:
Yes, it is. I have added a comment. That method is never called. It is there to satisfy the Key interface which is required to be able to write the metadata to the Java keystore.